src/Shared/WeakStringCache.cs

```diff
@@ -0,0 +1,355 @@
+        public string GetOrCreateEntry<T>(T internable, out bool cacheHit) where T : IInternable
+        {
+            int hashCode = GetInternableHashCode(internable);
+
+            StringBucket bucket;
+            string result;
+            lock (_stringsByHashCode)
+            {
+                if (!_stringsByHashCode.TryGetValue(hashCode, out bucket))
+                {
+                    bucket = new StringBucket();
+                }
+                result = bucket.GetString(internable);
+                if (result != null)
+                {
+                    cacheHit = true;
+                    return result;
+                }
+            }
+
+            // We don't have the string in the dictionary - create it.
+            result = internable.ExpensiveConvertToString();
+
+            lock (_stringsByHashCode)
+            {
+                bucket.AddString(result);
+
+                // Prevent the dictionary from growing forever with buckets whose underlying GC handles
+                // don't reference any live strings anymore.
+                if (_stringsByHashCode.Count >= _capacity)
+                {
+                    // Get rid of unused buckets.
+                    Scavenge();
+                    // And do this again if the number of buckets reaches double the current after-scavenge number.
+                    _capacity = _stringsByHashCode.Count * 2;
+                }
+                _stringsByHashCode[hashCode] = bucket;
+            }
+            cacheHit = false;
+            return result;
+        }
+
+        /// <summary>
+        /// Implements the simple yet very decently performing djb2 hash function.
+        /// </summary>
+        /// <param name="internable">The internable to compute the hash code for.</param>
+        /// <returns>The 32-bit hash code.</returns>
+        internal static int GetInternableHashCode<T>(T internable) where T : IInternable
+        {
+            int hashCode = 5381;
+            for (int i = 0; i < internable.Length; i++)
+            {
+                unchecked
+                {
+                    hashCode = hashCode * 33 + internable[i];
+                }
+            }
+            return hashCode;
+        }
+
+        /// <summary>
+        /// Iterates over the cache and removes unused buckets, i.e. buckets that don't reference live strings.
+        /// This is expensive so try to call such that the cost is amortized to O(1) per GetOrCreateEntry() invocation.
+        /// Assumes exclusive access to the dictionary, i.e. the lock is taken.
+        /// </summary>
+        public void Scavenge()
+        {
+            List<int> keysToRemove = null;
+            foreach (KeyValuePair<int, StringBucket> entry in _stringsByHashCode)
+            {
+                if (!entry.Value.IsUsed)
+                {
+                    entry.Value.Free();
+                    keysToRemove ??= new List<int>();
```

Comment:
I'm guessing you put this in the loop so you wouldn't allocate unnecessarily if none of the entries are used. What if you walked through this with a while loop, just checking for not unused entries and stopped on the first one, then had a second while loop that continued to the end doing this logic after allocating? This is only one line, but it's more complex than normal.

Reply:
This pattern has been shamelessly stolen from `WeakValueDictionary`. It's just a simple `if` and while I can measure a difference if the loop is doing nothing but iterating over the dictionary, in reality it's going to be dominated by the collection and handle operations. I lean towards not complicating the code in this case. Interesting idea, though!